src/main/java/engineering/swat/watch/impl/jdk/JDKFileTreeWatch.java

```diff
@@ -0,0 +1,278 @@
+                    }
+                }
+
+                var event = new WatchEvent(kind, rootPath, relativePath);
+                logger.trace("Translated: {} to {}", jdkEvent, event);
+                return event;
+            }
+        };
+    }
+
+    /**
+     * Event handler that asynchronously (using {@link JDKBaseWatch#exec})
+     * updates the child watches according to the following rules: (a) when an
+     * overflow happens, the directory is rescanned, new child watches for
+     * created subdirectories are opened, existing child watches for deleted
+     * subdirectories are closed, and the overflow is propagated to each child
+     * watch; (b) when a subdirectory creation happens, a new child watch is
+     * opened for that subdirectory; (c) when a subdirectory deletion happens,
+     * an existing child watch is closed for that subdirectory.
+     */
+    private class AsyncChildWatchesUpdater implements BiConsumer<EventHandlingWatch, WatchEvent> {
+        @Override
+        public void accept(EventHandlingWatch watch, WatchEvent event) {
+            exec.execute(() -> {
+                switch (event.getKind()) {
+                    case OVERFLOW: acceptOverflow(); break;
+                    case CREATED: getFileNameAndThen(event, this::acceptCreated); break;
+                    case DELETED: getFileNameAndThen(event, this::acceptDeleted); break;
+                    case MODIFIED: break;
+                }
+            });
+        }
+
+        private void getFileNameAndThen(WatchEvent event, Consumer<Path> consumer) {
+            var child = event.getFileName();
+            if (child != null) {
+                consumer.accept(child);
+            } else {
+                logger.error("Could not get file name of event: {}", event);
+            }
+        }
+
+        private void acceptOverflow() {
+            openAndCloseChildWatches();
+            for (var childWatch : childWatches.values()) {
+                reportOverflowTo(childWatch);
+            }
+        }
+
+        private void acceptCreated(Path child) {
+            if (Files.isDirectory(path.resolve(child))) {
+                var childWatch = openChildWatch(child);
+                // Events in the newly created directory might have been missed
+                // between its creation and setting up its watch. So, generate
+                // an `OVERFLOW` event for the watch.
+                reportOverflowTo(childWatch);
+            }
+        }
+
+        private void acceptDeleted(Path child) {
+            tryCloseChildWatch(child);
+        }
+
+        private void reportOverflowTo(JDKFileTreeWatch childWatch) {
+            var overflow = new WatchEvent(WatchEvent.Kind.OVERFLOW,
+                childWatch.rootPath, childWatch.relativePathParent);
+            childWatch.handleEvent(overflow);
+        }
+    }
+
+    private void openAndCloseChildWatches() {
```

Comment:
nit: better name `syncWithFileSystem`? since now the name is saying what it does, but not why.